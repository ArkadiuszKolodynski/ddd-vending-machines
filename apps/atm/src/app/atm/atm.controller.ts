import { Body, Controller, Get, HttpCode, Param, ParseUUIDPipe, Post } from '@nestjs/common';
import { CommandBus, QueryBus } from '@nestjs/cqrs';
import { GrpcMethod } from '@nestjs/microservices';
import { LoadMoneyDto, Money } from '@vending-machines/shared';
import { AtmService } from './atm.service';
import { TakeMoneyCommand } from './commands/impl/take-money.command';
import { AtmDto } from './dto/atm.dto';
import { TakeMoneyDto } from './dto/take-money.dto';
import { GetAtmQuery } from './queries/impl/get-atm.query';

@Controller('atm')
export class AtmController {
  constructor(
    private readonly commandBus: CommandBus,
    private readonly queryBus: QueryBus,
    private readonly atmService: AtmService,
  ) {}

  @Get(':id')
  getAtm(@Param('id', new ParseUUIDPipe({ version: '4' })) id: string): Promise<AtmDto> {
    return this.queryBus.execute(new GetAtmQuery(id));
  }

  @HttpCode(200)
  @Post(':id/take-money')
  async takeMoney(
    @Body() requestDto: TakeMoneyDto,
    @Param('id', new ParseUUIDPipe({ version: '4' })) id: string,
  ): Promise<AtmDto> {
    await this.commandBus.execute(new TakeMoneyCommand(id, requestDto.amount));
    return this.queryBus.execute(new GetAtmQuery(id));
  }

  @GrpcMethod('AtmProtoService', 'LoadMoney')
  loadMoney(requestDto: LoadMoneyDto): Promise<void> {
    return this.atmService.loadMoney(requestDto.id, new Money(...requestDto.money));
  }
}
