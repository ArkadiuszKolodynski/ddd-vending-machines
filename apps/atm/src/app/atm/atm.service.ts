import { CreateRequestContext, MikroORM } from '@mikro-orm/core';
import { Injectable, NotFoundException } from '@nestjs/common';
import { EventPublisher } from '@nestjs/cqrs';
import { Money } from '@vending-machines/shared';
import { AtmRepository } from './atm.repository.interface';

@Injectable()
export class AtmService {
  constructor(
    private readonly orm: MikroORM, // MikroORM needed for @CreateRequestContext()
    private readonly atmRepository: AtmRepository,
    private readonly eventPublisher: EventPublisher,
  ) {}

  @CreateRequestContext()
  async loadMoney(id: string, money: Money): Promise<void> {
    const atm = this.eventPublisher.mergeObjectContext(await this.atmRepository.findOne(id));

    if (!atm) {
      throw new NotFoundException(`Atm with id ${id} not found`);
    }

    atm.loadMoney(money);
    await this.atmRepository.save(atm);
    atm.commit();
  }
}
