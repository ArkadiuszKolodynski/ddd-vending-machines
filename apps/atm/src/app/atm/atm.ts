import { AggregateRoot, InvalidOperationException, Money } from '@vending-machines/shared';
import Currency from 'currency.js';

export class Atm extends AggregateRoot {
  moneyInside: Money = Money.None;
  moneyCharged: Currency = new Currency(0);
  private readonly commissionRate = new Currency(0.01);

  loadMoney(money: Money): void {
    this.moneyInside = Money.add(this.moneyInside, money);
  }

  canTakeMoney(amount: Currency): string {
    if (amount <= new Currency(0)) {
      return 'Invalid amount';
    }

    if (this.moneyInside.amount.value < amount.value) {
      return 'Not enough money';
    }

    if (!this.moneyInside.canAllocate(amount)) {
      return 'Not enough change';
    }

    return String();
  }

  takeMoney(amount: Currency): void {
    const takeMoneyError = this.canTakeMoney(amount);
    if (takeMoneyError) {
      throw new InvalidOperationException(takeMoneyError);
    }

    const output = this.moneyInside.allocate(amount);
    this.moneyInside = Money.subtract(this.moneyInside, output);

    const amountWithCommission = this.calculateAmountWithCommission(amount);
    this.moneyCharged = amountWithCommission;
  }

  private calculateAmountWithCommission(amount: Currency): Currency {
    const commission = amount.value * this.commissionRate.value;
    const lessThanCent = ((commission * 100) % (this.commissionRate.value * 100)) / 100;
    if (lessThanCent > 0) {
      return amount.add(commission).subtract(lessThanCent).add(this.commissionRate);
    }
    return amount.add(commission);
  }
}
