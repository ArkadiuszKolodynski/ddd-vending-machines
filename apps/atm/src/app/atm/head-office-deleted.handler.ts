import { CreateRequestContext, MikroORM } from '@mikro-orm/core';
import { Logger } from '@nestjs/common';
import { EventPublisher, EventsHandler, IEventHandler } from '@nestjs/cqrs';
import { HeadOfficeDeletedEvent } from '@vending-machines/events';
import { AtmRepository } from './atm.repository.interface';

@EventsHandler(HeadOfficeDeletedEvent)
export class HeadOfficeDeletedHandler implements IEventHandler<HeadOfficeDeletedEvent> {
  constructor(
    private readonly orm: MikroORM, // MikroORM needed for @CreateRequestContext()
    private readonly atmRepository: AtmRepository,
    private readonly eventPublisher: EventPublisher,
  ) {}

  @CreateRequestContext()
  async handle(event: HeadOfficeDeletedEvent) {
    const existingAtm = await this.atmRepository.findOne(event.aggregateId);
    if (!existingAtm) {
      Logger.error(`Head office with id ${event.aggregateId} not found`, this.constructor.name);
      return;
    }

    const atm = this.eventPublisher.mergeObjectContext(existingAtm);
    atm.markAsDeleted();
    await this.atmRepository.delete(atm.id);
    atm.commit();
  }
}
