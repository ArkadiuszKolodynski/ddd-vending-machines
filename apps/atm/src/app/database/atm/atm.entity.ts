import { BaseEntity, Embedded, Entity, PrimaryKey, Property } from '@mikro-orm/core';
import { MoneyEmbeddable } from '@vending-machines/shared';

@Entity({ tableName: 'atm' })
export default class AtmEntity extends BaseEntity<AtmEntity, 'id'> {
  @PrimaryKey({ type: 'uuid' })
  id!: string;

  @Property({ type: 'decimal', precision: 18, scale: 2 })
  moneyCharged!: number;

  @Embedded(() => MoneyEmbeddable)
  money!: MoneyEmbeddable;
}
