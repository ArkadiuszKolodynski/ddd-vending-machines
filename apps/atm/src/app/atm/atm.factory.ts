import { Injectable } from '@nestjs/common';
import { EventPublisher } from '@nestjs/cqrs';
import { AtmCreatedEvent } from '@vending-machines/events';
import { Atm } from './atm';

@Injectable()
export class AtmFactory {
  constructor(private readonly eventPublisher: EventPublisher) {}

  create(id?: string): Atm {
    const atm = this.eventPublisher.mergeObjectContext(new Atm(id));
    atm.apply(new AtmCreatedEvent({ aggregateId: atm.id, aggregateType: atm.constructor.name, payload: {} }));
    return atm;
  }
}
