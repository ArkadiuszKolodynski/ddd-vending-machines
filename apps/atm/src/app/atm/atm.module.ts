import { Module } from '@nestjs/common';
import { CqrsModule } from '@nestjs/cqrs';
import { Atm } from './atm';
import { AtmController } from './atm.controller';
import { AtmRepository } from './atm.repository.interface';
import { AtmService } from './atm.service';
import { CommandHandlers } from './commands/handlers';
import { QueryHandlers } from './queries/handlers';

@Module({
  imports: [CqrsModule],
  controllers: [AtmController],
  providers: [
    ...CommandHandlers,
    ...QueryHandlers,
    AtmService,
    {
      provide: Atm,
      useFactory: async (atmRepository: AtmRepository) => await atmRepository.findOne(),
      inject: [AtmRepository],
    },
  ],
})
export class AtmModule {}
