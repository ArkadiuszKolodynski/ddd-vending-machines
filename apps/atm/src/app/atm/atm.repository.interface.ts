import { Atm } from './atm';

export interface AtmRepository {
  findOne(): Atm | null | Promise<Atm | null>;
  save(atm: Atm): void | Promise<void>;
}

export const AtmRepository = Symbol('AtmRepository');
