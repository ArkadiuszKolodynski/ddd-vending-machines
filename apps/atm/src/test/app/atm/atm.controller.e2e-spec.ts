import { HttpStatus, INestApplication, Module } from '@nestjs/common';
import { CommandBus, QueryBus } from '@nestjs/cqrs';
import { Test } from '@nestjs/testing';
import { ValidationProvider } from '@vending-machines/shared';
import request from 'supertest';
import { AtmRepository } from '../../../app/atm/atm.repository.interface';
import { AtmService } from '../../../app/atm/atm.service';

@Module({
  providers: [{ provide: AtmRepository, useValue: { findOne: jest.fn(), save: jest.fn() } }],
  exports: [AtmRepository],
})
class DatabaseModuleMock {}
jest.mock('../../../app/database/database.module', () => {
  return {
    DatabaseModule: DatabaseModuleMock,
  };
});

import { randomUUID } from 'crypto';
import { AtmController } from '../../../app/atm/atm.controller';

describe('AtmController - e2e', () => {
  const queryBusMock = { execute: jest.fn(async () => ({ moneyCharged: '$1.00', moneyInside: '$1.00' })) };
  const testEndpoint = (id: string) => `/atm/${id}`;

  let app: INestApplication;

  beforeAll(async () => {
    const module = await Test.createTestingModule({
      controllers: [AtmController],
      providers: [
        ValidationProvider,
        { provide: QueryBus, useValue: queryBusMock },
        { provide: CommandBus, useValue: { execute: jest.fn() } },
        { provide: AtmService, useValue: { loadMoney: jest.fn() } },
      ],
    }).compile();

    app = module.createNestApplication();
    await app.init();
  });

  afterAll(async () => {
    await app.close();
  });

  describe('GET /:id', () => {
    it('should return 200 OK', async () => {
      return request(app.getHttpServer())
        .get(testEndpoint(randomUUID()))
        .expect(HttpStatus.OK)
        .expect(await queryBusMock.execute());
    });

    it('should return 400 BAD REQUEST when id is not a valid uuid', async () => {
      const response = await request(app.getHttpServer())
        .get(testEndpoint('not-valid-uuid'))
        .expect(HttpStatus.BAD_REQUEST);

      expect(response.body.message).toMatchSnapshot();
    });
  });

  describe('POST /:id/take-money', () => {
    it('should return 200 OK', async () => {
      return request(app.getHttpServer())
        .post(`${testEndpoint(randomUUID())}/take-money`)
        .send({ amount: '20' })
        .expect(HttpStatus.OK)
        .expect(await queryBusMock.execute());
    });

    it('should return 400 BAD REQUEST when id is not a valid uuid', async () => {
      const response = await request(app.getHttpServer())
        .post(`${testEndpoint('not-valid-uuid')}/take-money`)
        .send({ amount: '20' })
        .expect(HttpStatus.BAD_REQUEST);

      expect(response.body.message).toMatchSnapshot();
    });

    it('should return 400 BAD REQUEST when payload is not a valid currency', async () => {
      const response = await request(app.getHttpServer())
        .post(`${testEndpoint(randomUUID())}/take-money`)
        .send({ amount: '20.001' })
        .expect(HttpStatus.BAD_REQUEST);

      expect(response.body.message).toMatchSnapshot();
    });
  });
});
