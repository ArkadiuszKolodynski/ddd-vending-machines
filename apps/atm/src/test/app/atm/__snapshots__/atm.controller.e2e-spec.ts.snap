// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtmController - e2e GET /:id should return 400 BAD REQUEST when id is not a valid uuid 1`] = `"Validation failed (uuid v 4 is expected)"`;

exports[`AtmController - e2e POST /:id/take-money should return 400 BAD REQUEST when id is not a valid uuid 1`] = `"Validation failed (uuid v 4 is expected)"`;

exports[`AtmController - e2e POST /:id/take-money should return 400 BAD REQUEST when payload is not a valid currency 1`] = `"amount must be a valid currency"`;
